File: java/Playground/src/chap1/Exo1.java
package chap1;

import static org.junit.Assert.*;

public class Exo1 {

	/*
	 * Implement an algorithm to determine
	 * if a string has all unique characters.
	 */
	
	/*
	 * First algorithm.
	 * Use an extra array to retain every character we've seen so far.
	 * Check if we've seen the character. If so, return true, stop. otherwise
	 * continue till to the end 
	 */
	public static boolean hasUniqueCharacters(String word) {
		if (word.length() >256)
			return false;
		char[] verifier = new char[word.length()];
		char[] wordArray = word.toCharArray();
		for(int i=0, wordLength = wordArray.length; i < wordLength; i++) {
			if(isAlreadySeen(wordArray[i], verifier)) {
				return false;
			} else {
				verifier[i] = wordArray[i];
			}
		}
		return true;
	}
	
	public static boolean hasUniqueCharacters2(String word) {
		boolean[] checker = new boolean[256];
		if (word.length() > 256)
			return false;
		for(int i=0, l= word.length(); i<l; i++) {
			if (checker[word.charAt(i)]) {
				return false;
			}
			checker[word.charAt(i)] = true;
		}
		
		return true;
	}
	
	private static boolean isAlreadySeen(char c, char[] word) {
		for (int i=0, l = word.length; i < l; i++) {
			if (word[i] == c)
				return true;
		}
		return false;
	}
}

File: javascript/chap1/exo1.js
var assert = require('assert');

function hasUniqueCharacters(word) {
  if(word.length >= 256)
    return false;
  var checker = [];

  for(var i=0, l= word.length; i<l; i++) {
    if (checker[word.charAt(i)]) {
      return false;
    } else {
      checker[word.charAt(i)] = true;
    }
  }

  return true;
}

describe('checks if a word has unique characters.', function() {
  it('It should return true if a word is made of unique characters and false otherwise', function() {
    assert.equal(hasUniqueCharacters('diokey'), true);
    assert.equal(hasUniqueCharacters('Olivier'), false);
    assert.equal(hasUniqueCharacters('Dusabimana'), false); 
  });
});
